openapi: Use the Spider add-on when available

kingthorin opened the pull request:
- CHANGELOG > Added change notes.
- Build file > Included spider extension and updated dependencies.
- Separated spider functionality into a static class which is now used by both spider parser implementations (core & add-on).

Part of zaproxy/zaproxy#3113

Signed-off-by: kingthorin <kingthorin@users.noreply.github.com>

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/spider/ExtensionOpenApiSpider.java:
```diff
@@ -0,0 +1,80 @@
+import org.zaproxy.addon.spider.ExtensionSpider2;
+import org.zaproxy.addon.spider.parser.SpiderParser;
+
+public class ExtensionOpenApiSpider extends ExtensionAdaptor {
```
Would be good to have some tests.

kingthorin replied:
I'll try to look at what the existing one does. Open to suggestions though.

thc202 replied:
They would be similar to e.g. `ExtensionNetworkUnitTest#shouldAddNetworkApiOnHook` and `shouldUnloadIfHandlingServerCerts`. (Could also have `shouldBeUnloadable`.)

kingthorin replied:
Okay it seems it' s going to take me a while to re-implement all the similar infrastructure for that.

kingthorin replied:
Even just doing `shouldBeUnloadable` requires figuring out how to handle loggers etc.

kingthorin replied:
I have no idea 😞 

I tried doing something like `HistoryPersisterUnitTest#shouldLogWarnIfUnableToPersistMessage` didn't work.

kingthorin replied:
It almost seems better to just remove the two logging statements than fight with it.

thc202 replied:
What's the problem with the logger?

kingthorin replied:
I'll get the stack trace.

thc202 replied:
I think I know that one, probably the Eclipse's classpath issue.

kingthorin replied:
<details>
<summary>Stack Trace</summary>

```bash
java.lang.NoSuchMethodError: 'java.util.Map org.apache.logging.log4j.util.PropertiesUtil.partitionOnCommonPrefixes(java.util.Properties, boolean)'
	at org.apache.logging.log4j.core.config.properties.PropertiesConfigurationBuilder.build(PropertiesConfigurationBuilder.java:176)
	at org.apache.logging.log4j.core.config.properties.PropertiesConfigurationFactory.getConfiguration(PropertiesConfigurationFactory.java:56)
	at org.apache.logging.log4j.core.config.properties.PropertiesConfigurationFactory.getConfiguration(PropertiesConfigurationFactory.java:35)
	at org.apache.logging.log4j.core.config.ConfigurationFactory$Factory.getConfiguration(ConfigurationFactory.java:559)
	at org.apache.logging.log4j.core.config.ConfigurationFactory$Factory.getConfiguration(ConfigurationFactory.java:479)
	at org.apache.logging.log4j.core.config.ConfigurationFactory.getConfiguration(ConfigurationFactory.java:325)
	at org.apache.logging.log4j.core.LoggerContext.reconfigure(LoggerContext.java:690)
	at org.apache.logging.log4j.core.LoggerContext.reconfigure(LoggerContext.java:711)
	at org.apache.logging.log4j.core.LoggerContext.start(LoggerContext.java:253)
	at org.apache.logging.log4j.core.impl.Log4jContextFactory.getContext(Log4jContextFactory.java:155)
	at org.apache.logging.log4j.core.impl.Log4jContextFactory.getContext(Log4jContextFactory.java:47)
	at org.apache.logging.log4j.LogManager.getContext(LogManager.java:196)
	at org.apache.logging.log4j.LogManager.getLogger(LogManager.java:599)
	at org.zaproxy.zap.extension.openapi.spider.ExtensionOpenApiSpider.<clinit>(ExtensionOpenApiSpider.java:39)
	at org.zaproxy.zap.extension.openapi.spider.ExtensionOpenApiSpiderUnitTest.shouldBeUnloadable(ExtensionOpenApiSpiderUnitTest.java:33)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:214)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:210)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:95)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:91)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:60)
	at org.eclipse.jdt.internal.junit5.runner.JUnit5TestReference.run(JUnit5TestReference.java:98)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)

```

</details>

Seems like it, if I run from CLI it's fine.

kingthorin replied:
Okay so that addresses `shouldBeUnloadable`, I'll still have to figure out the infra for the others.

kingthorin replied:
Should the other tests you mentioned really be here or in the spider add-on?

thc202 replied:
Here, we want to verify that the parser is added/removed when the extension is hooked/unloaded.

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/spider/ExtensionOpenApiSpider.java:
```diff
@@ -0,0 +1,80 @@
+public class ExtensionOpenApiSpider extends ExtensionAdaptor {
+
+    public static final String NAME = "ExtensionOpenApiSpider";
+    private static final Logger LOG = LogManager.getLogger(ExtensionOpenApiSpider.class);
```
`LOGGER`

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/spider/ExtensionOpenApiSpider.java:
```diff
@@ -0,0 +1,80 @@
+
+    @Override
+    public void hook(ExtensionHook extensionHook) {
+        /* Custom spider is added in order to explore Open API definitions. */
```
Remove.

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/spider/ExtensionOpenApiSpider.java:
```diff
@@ -0,0 +1,80 @@
+        ExtensionSpider2 spider =
+                Control.getSingleton().getExtensionLoader().getExtension(ExtensionSpider2.class);
+        customSpider = new OpenApiSpider();
+        if (spider != null) {
```
This should not be needed, the extension should depend on `ExtensionSpider2` (and `ExtensionOpenApi`).

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/spider/OpenApiSpiderFunctionality.java:
```diff
@@ -0,0 +1,91 @@
+
+    private static final Logger LOGGER = LogManager.getLogger(OpenApiSpiderFunctionality.class);
+    private static Requestor requestor;
+    private static ValueGenerator valGen = null;
```
This should be provided by the parsers, each should get it from the corresponding spider extension.

thc202 commented:
The typo is also in the commit message.

kingthorin commented:
I believe the LGTM thing is an FP and the null guard there can likely go away.

thc202 commented:
The extension can be null there, when disabled.

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/spider/OpenApiSpiderFunctionality.java:
```diff
@@ -0,0 +1,87 @@
+public class OpenApiSpiderFunctionality {
+
+    private static final Logger LOGGER = LogManager.getLogger(OpenApiSpiderFunctionality.class);
+    private static Requestor requestor;
```
This is no longer being initialised, worth adding a test to ensure it's working properly.

kingthorin replied:
I re-added the initialization. Once the current test challenges are resolved, I can look into a test for this.

I looked quickly at just testing the constructor but there doesn't seem to be a way to `verify` the assignments within the constructor.

thc202 replied:
Right, you would have to use it (i.e. call `parseResource`, it will be like the other "import" tests so can be based on those).

kingthorin replied:
I guess I shouldn't have tried this (PR), I still can't sort this out.

thc202 replied:
This should do it:
```java
package org.zaproxy.zap.extension.openapi.spider;

import static fi.iki.elonen.NanoHTTPD.newFixedLengthResponse;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.contains;
import static org.mockito.Mockito.mock;

import fi.iki.elonen.NanoHTTPD.IHTTPSession;
import fi.iki.elonen.NanoHTTPD.Response;
import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.parosproxy.paros.network.HttpMessage;
import org.zaproxy.zap.extension.openapi.AbstractServerTest;
import org.zaproxy.zap.model.ValueGenerator;
import org.zaproxy.zap.testutils.NanoServerHandler;

/** Unit test for {@link OpenApiSpiderFunctionality}. */
class OpenApiSpiderFunctionalityUnitTest extends AbstractServerTest {

    private ValueGenerator valueGenerator;
    private OpenApiSpiderFunctionality spider;

    @BeforeEach
    void setupSpider() {
        valueGenerator = mock(ValueGenerator.class);
        spider = new OpenApiSpiderFunctionality(valueGenerator);
    }

    @Test
    void shouldParseResource() throws Exception {
        // Given
        List<String> accessedUris = new ArrayList<>();
        this.nano.addHandler(
                new NanoServerHandler("") {
                    @Override
                    protected Response serve(IHTTPSession session) {
                        accessedUris.add(session.getUri());
                        return newFixedLengthResponse("");
                    }
                });
        HttpMessage message = getHttpMessage("");
        message.setResponseBody(
                "openapi: 3.0.0\n"
                        + "servers:\n"
                        + "  - url: http://localhost:"
                        + nano.getListeningPort()
                        + "\n"
                        + "paths:\n"
                        + "  /path:\n"
                        + "    get:\n"
                        + "      responses:\n"
                        + "        200:\n"
                        + "          content:\n"
                        + "            application/json: {}");
        // When
        spider.parseResource(message, null, 0);
        // Then
        assertThat(accessedUris, contains("/path"));
    }
}
```

kingthorin replied:
TYVM @thc202

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/spider/OpenApiSpiderFunctionality.java:
```diff
@@ -0,0 +1,87 @@
+public class OpenApiSpiderFunctionality {
+
+    private static final Logger LOGGER = LogManager.getLogger(OpenApiSpiderFunctionality.class);
+    private static Requestor requestor;
+    private static ValueGenerator valGen = null;
```
Should be an instance variable.

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/spider/OpenApiSpiderFunctionality.java:
```diff
@@ -0,0 +1,87 @@
+        return false;
+    }
+
+    private static ValueGenerator getValueGenerator() {
+        return valGen;
+    }
```
Not really needed.

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/spider/OpenApiSpider.java:
```diff
@@ -0,0 +1,46 @@
+
+public class OpenApiSpider extends SpiderParser {
+
+    private ValueGenerator valGen = null;
```
This is not needed.

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/spider/OpenApiSpider.java:
```diff
@@ -0,0 +1,46 @@
+public class OpenApiSpider extends SpiderParser {
+
+    private ValueGenerator valGen = null;
+    private OpenApiSpiderFunctionality func = null;
```
Remove the null initialisation.

kingthorin commented:
I'm not sure what to do about the failure. I encountered it on my dev box as well but it went away if I did a "Build Project" then ran the gradle `test` task.

kingthorin commented:
I'm not sure what to do about the test failure. It seems to have something to do with Model being set in the new tests. I'm not sure how to unset or reset it afterwards. I tried doing an aftereach nulling it but that didn't help.

I did encounter something similar on my dev box but there doing a "build project" before doing a "gradle test" seemed to resolve things.

thc202 commented on addOns/openapi/src/test/java/org/zaproxy/zap/extension/openapi/spider/ExtensionOpenApiSpiderUnitTest.java:
```diff
@@ -0,0 +1,80 @@
+
+class ExtensionOpenApiSpiderUnitTest extends TestUtils {
+
+    private Model model;
+    private ExtensionLoader extensionLoader;
+    private ExtensionOpenApiSpider extension;
+
+    @BeforeEach
+    void setUp() {
+        extension = new ExtensionOpenApiSpider();
+        model = mock(Model.class, withSettings().lenient());
+        Model.setSingletonForTesting(model);
+
+        extensionLoader = mock(ExtensionLoader.class, withSettings().lenient());
+        Control.initSingletonForTesting(model, extensionLoader);
+    }
```
This should address the tests failing.

```suggestion
    private ExtensionLoader extensionLoader;
    private ExtensionOpenApiSpider extension;

    @BeforeEach
    void setUp() {
        extension = new ExtensionOpenApiSpider();

        extensionLoader = mock(ExtensionLoader.class, withSettings().lenient());
        Control.initSingletonForTesting(Model.getSingleton(), extensionLoader);
    }
```

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/spider/OpenApiSpiderFunctionality.java:
```diff
@@ -0,0 +1,87 @@
+
+    private static final Logger LOGGER = LogManager.getLogger(OpenApiSpiderFunctionality.class);
+    private Requestor requestor;
+    private ValueGenerator valGen = null;
```
The null initialisation can be removed.

kingthorin commented:
Thanks for the test fix, I'll get back to that other part this evening.

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/spider/ExtensionOpenApiSpider.java:
```diff
@@ -0,0 +1,87 @@
+    public void hook(ExtensionHook extensionHook) {
+        ExtensionSpider2 spider =
+                Control.getSingleton().getExtensionLoader().getExtension(ExtensionSpider2.class);
+        customSpider = new OpenApiSpider(spider.getValueGenerator());
```
This need to be changed, the value generator is no longer being set to the spider.

kingthorin replied:
Can I just get ExtensionFormHandler (similar to L52/53), then use formHandler.getValueGenerator() on 54?

thc202 replied:
No, it needs to be like what was done in the soap add-on, have another extension getting/setting the value generator from the form handler extension (and supply it to the parser through the main extension).

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/spider/OpenApiSpiderFunctionality.java:
```diff
@@ -0,0 +1,87 @@
+
+    private static final Logger LOGGER = LogManager.getLogger(OpenApiSpiderFunctionality.class);
+    private Requestor requestor;
+    private ValueGenerator valGen;
```
`Supplier<ValueGenerator>`

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/spider/OpenApiSpiderFunctionality.java:
```diff
@@ -0,0 +1,87 @@
+                            null,
+                            message.getRequestHeader().getURI().toString(),
+                            message.getResponseBody().toString(),
+                            valGen);
```
`valGen.get()`

kingthorin commented:
Updated:
        modified:   addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/OpenApiSpider.java
        modified:   addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/spider/OpenApiSpider.java
        modified:   addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/spider/OpenApiSpiderFunctionality.java
to use `Supplier<ValueGenerator>`

thc202 commented on addOns/openapi/src/main/java/org/zaproxy/zap/extension/openapi/OpenApiSpider.java:
```diff
@@ -19,81 +19,30 @@
-    private static final Logger log = LogManager.getLogger(OpenApiSpider.class);
-    private Requestor requestor;
-    private ValueGenerator valGen = null;
+    private Supplier<ValueGenerator> valGenSupplier = null;
```
This could be removed, not used (just passed to the other class).

kingthorin commented:
Done

thc202 commented:
Thank you!
